test/alphabet/alphabet_test.cpp

```diff
@@ -0,0 +1,337 @@
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ==========================================================================
+
+#include <gtest/gtest.h>
+#include <sstream>
+
+#include <seqan3/alphabet.hpp>
+
+using namespace seqan3;
+
+template <typename T>
+class alphabet : public ::testing::Test
+{};
+
+// add all alphabets here
+using alphabet_types = ::testing::Types<dna4, dna5, rna4, rna5, nucl16,
+                                        /*aa27,*/
+                                        /*gap, gapped<nucl16>, */
+                                        illumina18, dna4q>;
+
+TYPED_TEST_CASE(alphabet, alphabet_types);
+
+TYPED_TEST(alphabet, default_value_constructor)
+{
+    TypeParam t1;
+    TypeParam t2{};
+}
+
+TYPED_TEST(alphabet, copy_constructor)
+{
+    TypeParam t1;
+    TypeParam t2{t1};
+    TypeParam t3(t1);
+    EXPECT_EQ(t1, t2);
+    EXPECT_EQ(t2, t3);
+}
+
+TYPED_TEST(alphabet, move_constructor)
+{
+    TypeParam t0;
+    TypeParam t1{t0};
+
+    TypeParam t2{std::move(t1)};
+    EXPECT_EQ(t2, t0);
+    TypeParam t3(std::move(t2));
+    EXPECT_EQ(t3, t0);
+}
+
+TYPED_TEST(alphabet, copy_assignment)
+{
+    TypeParam t1;
+    TypeParam t2;
+    t2 = t1;
+    EXPECT_EQ(t1, t2);
+}
+
+// move assignment
+TYPED_TEST(alphabet, move_assignment)
+{
+    TypeParam t0;
+    TypeParam t1{t0};
+    TypeParam t2;
+    TypeParam t3;
+    t2 = std::move(t1);
+    EXPECT_EQ(t2, t0);
+    t3 = std::move(t2);
+    EXPECT_EQ(t3, t0);
+}
+
+// swap
+TYPED_TEST(alphabet, swap)
+{
+    TypeParam t0;
+    TypeParam t1{t0};
+    TypeParam t2{};
+    TypeParam t3{};
+
+    std::swap(t1, t2);
+    EXPECT_EQ(t2, t0);
+    EXPECT_EQ(t1, t3);
+}
+
+TYPED_TEST(alphabet, alphabet_size)
+{
+    EXPECT_GT(alphabet_size_v<TypeParam>, 0);
+}
```

Comment:
> this is as a test a little bit weak, but you can't assume more.

Yep.

> Maybe you can pair each size with the type.

No, I really want generic test cases here, they should not be hardcoded.